src/never.rs

```diff
@@ -0,0 +1,91 @@
+use {Future, Poll};
+
+/// A future which is never resolved.
+///
+/// This future can be created with the `empty` function.
+#[derive(Copy, Clone)]
+pub struct Never {}
+
+/// Creates a future which never resolves, representing a computation that never
+/// finishes.
+///
+/// The returned future will never resolve with a success but is still
+/// susceptible to cancellation. That is, if a callback is scheduled on the
+/// returned future, it is only run once the future is dropped (canceled).
+pub fn never<T, E>() -> impl Future<Item = T, Error = E> + Send + 'static {
```

Comment:
You're correct, it doesn't require `!`. The function I was trying to write would default any unresolved type parameters to `!`, but that doesn't work because of the removal of default type parameters on functions.

It occurs to me that all I'm really looking for here is to have `Never`'s types default to `!`, which we can do with a default type parameter on a struct which has `PhantomData` fields like you had before. I'll try out that approach tonight and see if it makes sense.